In [2]:
import pandas as pd 
import os, gzip 
import re
from funciones_aux.desempaquetado  import *


**Desanidamos los archivos user_reviews y users_items. Para solucionar esto usamos el modulo desempaquedo**

# ETL: Users_items

## Extracción de users_items

In [116]:
path_data_json =  os.path.join('..','data','raw','users_items.json.gz')
path_data_csv =  os.path.join('..','data','clear','users_items.csv.gz')

# user_items = desempaquetado_users_items(path_data_json, path_data_csv)
user_items = pd.read_csv(path_data_csv, compression='gzip')

## Transformación de users_items

### Eliminación de datos nulos
- Dado que no existe información suficiente para inputar datos en users_items procederemos a eliminar estas filas nulas

In [117]:
user_items.dropna(inplace=True)

## Carga de data de users_items

In [118]:
path_data_csv =  os.path.join('..','data','clear','users_items.csv.gz')

user_items.to_csv(path_data_csv,index = False)


**Observación**

Inspeccionando users_items, con la funcón **desempaquetado_users_items** nos quedamos únicamente con las columnas que eran necesarias, la columna item_name la eliminamos ya que se puede obtener de steam games por medio de item_id.

# ETL User_Reviews

## Extraccion de user_reviews

In [3]:
path_data_json =  os.path.join('..','data','raw','user_reviews.json.gz')

reviews = desempaquetado_reviews(path_data_json)


## Tranformacion de user_reviews

### Eliminación de datos nulos
- Existen 28 usuarios que no tienen información, por lo que vamos a proceder a eliminar

In [4]:
reviews[reviews.isnull().any(axis=1)]

user_id funny posted last_edited item_id helpful recommend  \
137                gdxsd   NaN    NaN         NaN     NaN     NaN       NaN   
177    76561198094224872   NaN    NaN         NaN     NaN     NaN       NaN   
2559   76561198021575394   NaN    NaN         NaN     NaN     NaN       NaN   
10080            cmuir37   NaN    NaN         NaN     NaN     NaN       NaN   
13767          Jaysteeny   NaN    NaN         NaN     NaN     NaN       NaN   
15493             ML8989   NaN    NaN         NaN     NaN     NaN       NaN   
19184  76561198079215291   NaN    NaN         NaN     NaN     NaN       NaN   
20223  76561198079342142   NaN    NaN         NaN     NaN     NaN       NaN   
25056  76561198061996985   NaN    NaN         NaN     NaN     NaN       NaN   
26257  76561198108286351   NaN    NaN         NaN     NaN     NaN       NaN   
26674     CallTripleZero   NaN    NaN         NaN     NaN     NaN       NaN   
28444       Priceless612   NaN    NaN         NaN     NaN     NaN       NaN   
28452      diabolical666   NaN    NaN         NaN     NaN     NaN       NaN   
28975         DrScottnik   NaN    NaN         NaN     NaN     NaN       NaN   
31955  76561198110176420   NaN    NaN         NaN     NaN     NaN       NaN   
32819         jojoPL1987   NaN    NaN         NaN     NaN     NaN       NaN   
33785        Kylee_Kylie   NaN    NaN         NaN     NaN     NaN       NaN   
39279             _samie   NaN    NaN         NaN     NaN     NaN       NaN   
39357   FifthExclamation   NaN    NaN         NaN     NaN     NaN       NaN   
39420             sbenji   NaN    NaN         NaN     NaN     NaN       NaN   
40661     SpatialWheel14   NaN    NaN         NaN     NaN     NaN       NaN   
41777  76561198065238937   NaN    NaN         NaN     NaN     NaN       NaN   
44384  76561198071075080   NaN    NaN         NaN     NaN     NaN       NaN   
50718  76561198063489150   NaN    NaN         NaN     NaN     NaN       NaN   
53323   thecommunistbear   NaN    NaN         NaN     NaN     NaN       NaN   
55425  76561198085883839   NaN    NaN         NaN     NaN     NaN       NaN   
57578  76561198100396147   NaN    NaN         NaN     NaN     NaN       NaN   
57766  76561198102423654   NaN    NaN         NaN     NaN     NaN       NaN   

      review  
137      NaN  
177      NaN  
2559     NaN  
10080    NaN  
13767    NaN  
15493    NaN  
19184    NaN  
20223    NaN  
25056    NaN  
26257    NaN  
26674    NaN  
28444    NaN  
28452    NaN  
28975    NaN  
31955    NaN  
32819    NaN  
33785    NaN  
39279    NaN  
39357    NaN  
39420    NaN  
40661    NaN  
41777    NaN  
44384    NaN  
50718    NaN  
53323    NaN  
55425    NaN  
57578    NaN  
57766    NaN

In [5]:
reviews.dropna(inplace=True)

### Imputación de columna funny
- Esta columna contiene informacón de si a alguien le pareció divertido el comentario, vamos a extraer el número de personas a las que el comentario les pareció divertido. 

In [6]:

def num_is_funny(row):
  """ 
  Args:
      row (str): texto que quiero extraer el número.

  Returns:
      int : retorna el número de personas que le pareció divertido
  """
  
  match = re.search(r'(\d+) people found', row)
  if match :
    return int(match.group(1))
  return (0)

reviews['funny_count'] = reviews['funny'].apply(num_is_funny)

### Imputación columna helpful

In [7]:
pd.DataFrame(reviews['helpful'].unique(), columns=['Helpfull Unicos']).sample(10)

Helpfull Unicos
153    83 of 106 people (78%) found this review helpful
555    35 of 102 people (34%) found this review helpful
613   145 of 207 people (70%) found this review helpful
1201     8 of 22 people (36%) found this review helpful
303      8 of 17 people (47%) found this review helpful
741   155 of 179 people (87%) found this review helpful
490    41 of 41 people (100%) found this review helpful
616   493 of 618 people (80%) found this review helpful
1025  129 of 160 people (81%) found this review helpful
517       1 of 29 people (3%) found this review helpful

In [8]:
reviews[reviews['helpful'] == '17 of 73 people (23%) found this review helpful']

user_id                             funny  \
17797  OneEyedWolf  4 people found this review funny   

                         posted last_edited item_id  \
17797  Posted January 10, 2015.              340460   

                                               helpful recommend  \
17797  17 of 73 people (23%) found this review helpful      True   

                                                  review  funny_count  
17797  Well.. this game is better than doing nothing,...            4

In [9]:
## De las personas que comentaron, calculamos el porcentaje de personas que si ayudó.add()
## Para no desbalancear la data, las personas que no tenian información las imputaremos con 
## 50 % es decir la misma probabilidad que le sirva o no el comentario.

reviews['percentage_helpful']  = reviews['helpful'].str.extract(r'\((\d+)%\)').astype(float)/100
reviews['percentage_helpful'] = reviews['percentage_helpful'].fillna(0.5)

### Imputación columna recommend
- Transformaremos los datos de recomendaciones en 0 == no recomendado  y 1 == recomendado
- Tambien les daremos el tipo de categorico a la columna

In [10]:
reviews['recommend'] = reviews['recommend'].astype('int').astype(('category'))

### Imputación columna posted
- Vamos a extraer el año que se posteó y el año de la última actualizacionó

In [11]:
reviews['posted'].unique()

reviews['posted'] = reviews['posted'].str.replace('.', '')

reviews['year_posted'] = reviews['posted'].str.extract(r'(\d{4})')
reviews['last_edited'] = reviews['posted'].str.extract(r'(\d{4})')

reviews.rename(columns={'last_edited': 'year_last_edited'},inplace=True)


In [12]:
pd.DataFrame(reviews.isna().sum(),columns=['Nulos'])

Nulos
user_id                 0
funny                   0
posted                  0
year_last_edited    10119
item_id                 0
helpful                 0
recommend               0
review                  0
funny_count             0
percentage_helpful      0
year_posted         10119

#### Nota:
- Podemos observar que existen 10119 comentarios que no tenian la fecha del año, lamentablemente no tenemos información suficiente para inputar estos registros por lo que vamos proceder a eliminar.

- Notemos que vamos a eliminar el 10% de la data. Este problema se podria solución conocimento un poco más el conjunto de datos, pero por ahora por practicidad y eficiencia en la consulta, procedere a eliminarlas.

In [13]:
reviews.dropna(subset=['posted','year_last_edited'], inplace=True)

In [14]:
columnas = ['item_id', 'year_last_edited','year_posted', 'recommend', 'funny_count', 'percentage_helpful', 'review']

reviews_clear = reviews[columnas]

## Carga de data de user_reviews

In [15]:
path_data_csv =  os.path.join('..','data','clear','user_reviews.csv.gz')
reviews_clear.to_csv(path_data_csv)

# ETL : Steam_games

## Extración de Steam_games

In [16]:
path_data =  os.path.join('..','data','raw','steam_games.json.gz')

steam_games = pd.read_json(path_data, compression='gzip',lines=True)
 

## Transformación de Steam_games

In [19]:
# Eliminacion de filas que todos sus datos son nulos.
steam_games.dropna(how='all',inplace=True)

# Datos duplicados
steam_games[steam_games['id'].duplicated(keep=False)]

publisher  \
88384                                                None   
102204                                 Bethesda Softworks   
102883                                 Bethesda Softworks   
119271  Warner Bros. Interactive Entertainment, Feral ...   

                     genres                                        app_name  \
88384                  None                                            None   
102204             [Action]                Wolfenstein II: The New Colossus   
102883             [Action]                Wolfenstein II: The New Colossus   
119271  [Action, Adventure]  Batman: Arkham City - Game of the Year Edition   

                                                 title  \
88384                                             None   
102204                Wolfenstein II: The New Colossus   
102883                Wolfenstein II: The New Colossus   
119271  Batman: Arkham City - Game of the Year Edition   

                                                      url release_date  \
88384                      http://store.steampowered.com/         None   
102204          http://store.steampowered.com/app/612880/   2017-10-26   
102883  http://store.steampowered.com/app/612880/Wolfe...   2017-10-26   
119271           http://store.steampowered.com/app/200260   2012-09-07   

                                                     tags  \
88384                                                None   
102204  [Action, FPS, Gore, Violent, Alternate History...   
102883  [Action, FPS, Gore, Violent, Alternate History...   
119271  [Action, Open World, Batman, Adventure, Stealt...   

                                              reviews_url  \
88384                                                None   
102204  http://steamcommunity.com/app/612880/reviews/?...   
102883  http://steamcommunity.com/app/612880/reviews/?...   
119271                                               None   

                                                    specs  price  \
88384                                                None  19.99   
102204  [Single-player, Steam Achievements, Full contr...  59.99   
102883  [Single-player, Steam Achievements, Full contr...  59.99   
119271  [Single-player, Steam Achievements, Steam Trad...  19.99   

        early_access        id                                   developer  
88384            0.0       NaN                                        None  
102204           0.0  612880.0                               Machine Games  
102883           0.0  612880.0                               Machine Games  
119271           0.0       NaN  Rocksteady Studios,Feral Interactive (Mac)

**Nota** inspeccionando los datos, vamos a eliminar las siguietes filas.
- 88384 :	en esta fila no encontramos niguna información

- 102204 y 119271 se encuentran repetida 

In [20]:
# Eliminación de datos duplicados

index_duplicados_drop = [88384,102204,119271]
steam_games.drop(index=index_duplicados_drop, axis=1, inplace=True)

In [21]:
# Transforma una lista de elementos en un texto separado por coma
def swap_to_text(x : str):
  """Convierte una lista de strings en un solo texto separado por comas y espacios."""
  if x is not None:
      return ', '.join(x)
  return x


steam_games['genres'] = steam_games['genres'].apply(swap_to_text)
steam_games['tags'] = steam_games['tags'].apply(swap_to_text)
steam_games['specs'] = steam_games['specs'].apply(swap_to_text)

### Tratamiento de datos nulos

In [22]:
pd.DataFrame(steam_games.isna().sum(),columns=['Nulls'])

Nulls
publisher      8051
genres         3282
app_name          1
title          2049
url               0
release_date   2066
tags            162
reviews_url       0
specs           669
price          1377
early_access      0
id                0
developer      3298

#### Analisis puntual de nulos

In [23]:
# Eliminaremos las filas con app_name  nulo 
indices_a_borrar = steam_games[steam_games.app_name.isna()].index
steam_games.drop(index=indices_a_borrar, inplace=True)

# Rellenamos los titulos vacios con la columna app_name
steam_games['title'] = steam_games['app_name'].fillna(steam_games['title'])

# Completar los generos vacios con las etiquetas de los Tags
steam_games['genres'] = steam_games['tags'].fillna(steam_games['genres'])

# Si el dato en publisher se encuentrar nulo, entonces vamos a asumir que fue el desarrollador el publicador
steam_games['publisher'].fillna(steam_games['developer'],inplace=True)

# Si el dato en desarrollador se encuentrar nulo, entonces vamos a asumir que fue el publicador
steam_games['developer'].fillna(steam_games['publisher'],inplace=True)

# Si no hay datos del desarrollador y el publicador, entonces precederemos llenar 'otros'
steam_games['publisher'].fillna('Otros', inplace=True)
steam_games['developer'].fillna('Otros', inplace=True)

#### Capitalización de texto

In [24]:
# Capitalizacion de texto
columns_capitalize = ['developer', 'publisher', 'app_name','title']

for column in columns_capitalize:
  steam_games[column] = steam_games[column].str.capitalize()
  
# Eliminacion de espacios en blanco
steam_games[columns_capitalize] = steam_games[columns_capitalize].apply(lambda x: x.str.strip())

#### Tratamiento de fechas

In [25]:
# Cambiamos a formato fecha columna release_date y pusumos nulos los que no tenian un formato adecuado para poder establecer fecha

steam_games['release_date'] = pd.to_datetime(steam_games['release_date'], format='%Y-%m-%d', errors='coerce')

In [26]:
# Interpolamos fechas nulas usando interpolate. Este método lo he utilzado porque evita que cambie la distrbución de mis datos 
steam_games['release_date'] = steam_games['release_date'].interpolate()
steam_games['release_date'] = pd.to_datetime(steam_games['release_date'], format='%Y-%m-%d', errors='coerce')

In [27]:
# Crearemos las nuevas columnas 'release_year' y 'release_month'
steam_games['release_year'] = steam_games['release_date'].dt.year
steam_games['release_month'] = steam_games['release_date'].dt.month
steam_games.drop(columns='release_date',inplace=True)

### Analísis de columna Price

In [30]:
def filtro_reg_exp(reg_exp, df):
  "Filtra una columan por una expresion regular reg_exp "
  
  mask = df.str.contains(reg_exp, case=False, na=False)
  return df[mask]

In [31]:
# Encontremos los valores de precio que son de tipo str
filtro_reg_exp(r'\D+' ,steam_games['price']).unique()

array(['Free To Play', 'Free to Play', 'Free', 'Free Demo',
       'Play for Free!', 'Install Now', 'Play WARMACHINE: Tactics Demo',
       'Free Mod', 'Install Theme', 'Third-party', 'Play Now',
       'Free HITMAN™ Holiday Pack', 'Play the Demo',
       'Starting at $499.00', 'Starting at $449.00', 'Free to Try',
       'Free Movie', 'Free to Use'], dtype=object)

In [32]:
## Imputaremoe estos datos a mano, ya que son muy pocos.

input_price = {'Free To Play': 0, 
              'Free to Play': 0, 
              'Free': 0, 
              'Free Demo': 0,
              'Play for Free!': 0,
              'Install Now': 0, 
              'Play WARMACHINE: Tactics Demo': 0,
              'Free Mod': 0, 
              'Install Theme': 0, 
              'Third-party': 0,
              'Play Now': 0,
              'Free HITMAN™ Holiday Pack': 0, 
              'Play the Demo': 0,
              'Starting at $499.00': 499, 
              'Starting at $449.00': 449, 
              'Free to Try': 0,
              'Free Movie': 0, 
              'Free to Use': 0}

#Imputacion de precios
steam_games['price'] = steam_games['price'].replace(input_price)

In [34]:
steam_games.dropna(inplace= True)

## Carga de data de Steam_games

In [37]:
path_data_csv =  os.path.join('..','data','clear','steam_games.csv.gz')

steam_games.to_csv(path_data_csv, index=False)